packages/turbo/bin/turbo

```diff
@@ -1,10 +1,257 @@
+  const TURBO_BINARY_PATH = process.env.TURBO_BINARY_PATH;
+  if (TURBO_BINARY_PATH) {
+    if (!fs.existsSync(TURBO_BINARY_PATH)) {
+      console.warn(
+        `[turbo] Ignoring bad configuration: TURBO_BINARY_PATH=${TURBO_BINARY_PATH}`
+      );
+    } else {
+      return TURBO_BINARY_PATH;
+    }
+  }
+
+  const availablePlatforms = [
+    'darwin',
+    'linux',
+    'windows',
+  ];
+
+  const availableArchs = [
+    '64',
+    'arm64',
+  ];
+
+  // We need to figure out which binary to hand the user.
+  // The only place where the binary can be at this point is `require.resolve`-able
+  // relative to this package as it should be installed as an optional dependency.
+
+  const { platform, arch } = process;
+  const resolvedArch = arch === 'x64' ? '64' : arch;
+  const ext = platform === 'windows' ? '.exe' : '';
+
+  // Try all places in order until we get a hit.
+
+  // 1. The package which contains the binary we _should_ be running.
+  const correctBinary = availablePlatforms.includes(platform) && availableArchs.includes(resolvedArch) ? `turbo-${platform}-${resolvedArch}/bin/turbo${ext}` : null;
+  if (correctBinary !== null) {
+    try {
+      return require.resolve(`${correctBinary}`);
+    } catch (e) {}
+  }
+
+  // 2. Install the binary that they need just in time.
+  if (SHOULD_INSTALL && correctBinary !== null) {
+    console.warn('Turborepo did not find the correct binary for your platform.');
+    console.warn('We will attempt to install it now.');
+
+    try {
+      installUsingNPM();
+      const resolvedPath = require.resolve(`${correctBinary}`);
+      console.warn('Installation has succeeded.');
+      return resolvedPath;
+      } catch (e) {
+      console.warn('Installation has failed.');
+    }
+  }
+
+  // 3. Both Windows and macOS ARM boxes can run x64 binaries. Attempt to run under emulation.
+  const alternateBinary = (arch === "arm64" && ['darwin', 'windows'].includes(platform)) ? `turbo-${platform}-64/bin/turbo${ext}` : null;
+  if (SHOULD_ATTEMPT_EMULATED && alternateBinary !== null) {
+    try {
+      const resolvedPath = require.resolve(`${alternateBinary}`);
+      console.warn(`Turborepo detected that you're running:\n${platform} ${resolvedArch}.`);
+      console.warn(`We were not able to find the binary at:\n${correctBinary}`);
+      console.warn(`We found a possibly-compatible binary at:\n${alternateBinary}`);
+      console.warn(`We will attempt to run that binary.`);
+      return resolvedPath;
+    } catch (e) {}
+  }
+
+  // We are not going to run `turbo` this invocation.
+  // Let's give the best error message that we can.
+
+  // Possible error scenarios:
+  // - The user is on a platform/arch combination we do not support.
+  // - We somehow got detection wrong and never attempted to run the _actual_ `correctBinary` or `alternateBinary`.
+  // - The user doesn't have the correct packages installed for their platform.
+
+  // Explain our detection attempt:
```

Comment:
```
Turborepo failed to start.

Turborepo detected that you are running:
darwin arm64

***

We were not able to find the binary at:
turbo-darwin-arm64/bin/turbo

We looked for it at:
/Users/nathanhammond/repos/triage/test-npm-lockfile/node_modules/turbo/bin/node_modules
/Users/nathanhammond/repos/triage/test-npm-lockfile/node_modules/turbo/node_modules
/Users/nathanhammond/repos/triage/test-npm-lockfile/node_modules
/Users/nathanhammond/repos/triage/node_modules
/Users/nathanhammond/repos/node_modules
/Users/nathanhammond/node_modules
/Users/node_modules
/node_modules
/Users/nathanhammond/.node_modules
/Users/nathanhammond/.node_libraries
/Users/nathanhammond/.nvm/versions/node/v18.15.0/lib/node

***

Your platform (darwin) can sometimes run x86 under emulation.
We did not find a possibly-compatible binary at:
turbo-darwin-64/bin/turbo

We looked for it at:
/Users/nathanhammond/repos/triage/test-npm-lockfile/node_modules/turbo/bin/node_modules
/Users/nathanhammond/repos/triage/test-npm-lockfile/node_modules/turbo/node_modules
/Users/nathanhammond/repos/triage/test-npm-lockfile/node_modules
/Users/nathanhammond/repos/triage/node_modules
/Users/nathanhammond/repos/node_modules
/Users/nathanhammond/node_modules
/Users/node_modules
/node_modules
/Users/nathanhammond/.node_modules
/Users/nathanhammond/.node_libraries
/Users/nathanhammond/.nvm/versions/node/v18.15.0/lib/node

***

We did not find any binaries on this system.
This can happen if you run installation with the --no-optional flag.

***

Turborepo detected that your lockfile (/Users/nathanhammond/repos/triage/test-npm-lockfile/package-lock.json) does not enumerate all available platforms.
This is likely a consequence of an npm issue: https://github.com/npm/cli/issues/4828.

To resolve this issue for your repository, run:
npm install turbo@^1.10.12 --package-lock-only --save-dev && npm install

You will need to commit the updated lockfile.

***

If you believe this is an error, please include this message in your report.
```

Reply:
There is one other branch not represented in that output:

```
Turborepo checked to see if binaries for another platform are installed.
This typically indicates an error in sharing of pre-resolved node_modules across platforms.
One common reason for this is copying files to Docker.

We found these unnecessary binaries:
turbo-windows-64/bin/turbo.exe
```